sockeye/decoder.py

```diff
@@ -886,3 +908,384 @@ def _context_gate(self,
+    :param weight_normalization: Weight normalization.
+    :param embed_dropout: Dropout probability for target embeddings.
+    :param hidden_dropout: Dropout probability on next decoder hidden state.
+    """
+
+    def __init__(self,
+                 cnn_config: convolution.ConvolutionConfig,
+                 vocab_size: int,
+                 max_seq_len_target: int,
+                 num_embed: int,
+                 encoder_num_hidden: int,
+                 num_layers: int,
+                 positional_embedding_type: str,
+                 weight_tying: bool,
+                 weight_normalization: bool = False,
+                 embed_dropout: float = .0,
+                 hidden_dropout: float = .0) -> None:
+        super().__init__()
+        if embed_dropout > 0 and hidden_dropout > 0:
+            logger.warning("Setting cnn encoder dropout AND hidden dropout > 0 leads to "
+                           "two dropout layers on top of each other.")
+        self.cnn_config = cnn_config
+        self.vocab_size = vocab_size
+        self.max_seq_len_target = max_seq_len_target
+        self.num_embed = num_embed
+        self.encoder_num_hidden = encoder_num_hidden
+        self.num_layers = num_layers
+        self.positional_embedding_type = positional_embedding_type
+        self.weight_tying = weight_tying
+        self.weight_normalization = weight_normalization
+        self.embed_dropout = embed_dropout
+        self.hidden_dropout = hidden_dropout
+
+
+class ConvolutionalDecoder(Decoder):
+    """
+    Convolutional decoder similar to Gehring et al. 2017.
+
+    The decoder consists of an embedding layer, positional embeddings, and layers
+    of convolutional blocks with residual connections.
+
+    Notable differences to Gehring et al. 2017:
+     * Here the context vectors are created from the last encoder state (instead of using the last encoder state as the
+       key and the sum of the encoder state and the source embedding as the value)
+     * The encoder gradients are not scaled down by 1/(2 * num_attention_layers).
+     * Residual connections are not scaled down by math.sqrt(0.5).
+     * Attention is computed in the hidden dimension instead of the embedding dimension (removes need for training
+       several projection matrices)
+
+    :param config: Configuration for convolutional decoder.
+    :param embed_weight: Optionally use an existing embedding matrix instead of creating a new target embedding.
+    :param prefix: Name prefix for symbols of this decoder.
+    """
+
+    def __init__(self,
+                 config: ConvolutionalDecoderConfig,
+                 embed_weight: Optional[mx.sym.Symbol] = None,
+                 prefix: str = C.DECODER_PREFIX) -> None:
+        self.config = config
+        self.prefix = prefix
+
+        # TODO: potentially project the encoder hidden size to the decoder hidden size.
+        utils.check_condition(config.encoder_num_hidden == config.cnn_config.num_hidden,
```

Comment:
the dot product attention is done directly on the source embeddings, otherwise I'd need to add an additional projection. Maybe I should just add it though...